src/main/java/com/ezasm/gui/editor/Autocomplete.java

```diff
@@ -0,0 +1,139 @@
+package com.ezasm.gui.editor;
+
+import org.fife.ui.rsyntaxtextarea.RSyntaxTextArea;
+
+import java.awt.event.ActionEvent;
+import java.util.Collections;
+import java.util.List;
+
+import javax.swing.AbstractAction;
+import javax.swing.JTextField;
+import javax.swing.SwingUtilities;
+import javax.swing.event.DocumentEvent;
+import javax.swing.event.DocumentListener;
+import javax.swing.text.BadLocationException;
+import java.util.regex.*;
+
+public class Autocomplete implements DocumentListener {
+
+    private static enum Mode {
+        INSERT, COMPLETION
+    };
+
+    private RSyntaxTextArea textField;
+    private final List<String> keywords;
+    private Mode mode = Mode.INSERT;
+    private Pattern line_finder = Pattern.compile("[ \\t]*[_a-zA-Z][_a-zA-Z0-9]*:");
+
+    public Autocomplete(RSyntaxTextArea textField, List<String> keywords) {
+        this.textField = textField;
+        this.keywords = keywords;
+        Collections.sort(keywords);
+    }
+
+    @Override
+    public void changedUpdate(DocumentEvent ev) {
+
+    }
+
+    @Override
+    public void removeUpdate(DocumentEvent ev) {
+    }
+
+    @Override
+    public void insertUpdate(DocumentEvent ev) {
+        if (ev.getLength() != 1)
+            return;
+
+        int pos = ev.getOffset();
+        String content = null;
+        try {
+            content = textField.getText(0, pos + 1);
+        } catch (BadLocationException e) {
+            e.printStackTrace();
+        }
+
+        // Find where the word starts
+        int w;
+        for (w = pos; w >= 0; w--) {
+            if (!Character.isLetterOrDigit(content.charAt(w)) && content.charAt(w) != '_') {
+                break;
+            }
+        }
+
+        // Too few chars
+        if (pos - w < 2) {
+            return;
+        } else {
+            // only recheck keywords once we've typed 3 non-space characters
+            keywords.clear();
+            String s = textField.getText();
+            Matcher m = line_finder.matcher(s);
+            while (m.find()) {
```

Comment:
Rebuilding the whole list of keywords seems inefficient especially in larger files. Try this library instead?
https://github.com/bobbylight/AutoComplete